Starting architecture for using a reducer.

kshehata opened the pull request:
Hey Manuel! I've started going through the architecture for adding reducer support on this branch. Would you please take a look and let me know if this is heading in the right direction?

Here's what I've done so far:

1. Used the FoundBugException from the previous PRs for capturing state and checking if the bug reproduces.
2. Added an interface for Reducer support. Idea is that eventually that should be made more generic.
3. Moved all of the reducing code that was in Main into an example reducer file. The idea is to eventually replace that with something more robust.

I'm thinking that before merging this should probably be hidden behind a command line flag, but just wanted to get your feedback at this stage.

Thanks!

mrigger commented on src/sqlancer/Main.java:
```diff
@@ -124,6 +127,17 @@ private FileWriter getLogFileWriter() {
         }
 
+        private FileWriter getReducedWriter() {
+            // if (logFileWriter == null) {
```
perhaps you can remove the comment?

kshehata replied:
Removed.

mrigger commented on src/sqlancer/Main.java:
```diff
@@ -172,6 +186,17 @@ private void write(Loggable loggable) {
+            try {
+                logFileWriter.close();
+            } catch (IOException e) {
+                // TODO Auto-generated catch block
```
here, we could throw an `AssertionError` instead.

kshehata replied:
Removed.

mrigger commented on src/sqlancer/Main.java:
```diff
@@ -312,30 +345,30 @@ public void run() throws Exception {
+
+                } catch (FoundBugException e) {
+                    try {
+                        logger.getCurrentFileWriter().close();
```
This seems quite hacky (perhaps part of my initial prototype implementation), not sure if there is a better way to implement this functionality.

kshehata replied:
TBH, I don't understand enough of how Main works. This is taken from your original prototype, and I'm open to suggestions for how to do it better!

mrigger commented on src/sqlancer/Main.java:
```diff
@@ -490,6 +523,7 @@ private boolean run(MainOptions options, ExecutorService execService,
                     DBMSExecutor<?, ?, ?> executor = executorFactory.getDBMSExecutor(databaseName, r);
                     try {
                         executor.run();
+                        System.out.println("Executor completed!");
```
I think we should not check-in this line.

kshehata replied:
Sorry about that, removed!

mrigger commented on src/sqlancer/ReducerExample.java:
```diff
@@ -0,0 +1,120 @@
+            });
+        } while (observedChange);
+
+        for (String s : TOKENS) {
```
For an initial implementation, I'd suggest keeping it general and simple, by only removing complete statements. This might simplify the implementation.

kshehata replied:
Done.

mrigger reviewed (commented):
Thanks, Kareem! This PR definitely goes in the right direction.

I agree that it would be a good idea to hide things behind an option flag.

What do you think of a PR that adds the essential classes and interfaces (like `FoundBugException` or `Reducer`) and adopts the `FoundBugException` without actually changing any of the visible behavior, and then a follow-up PR that adopts the new classes and the reducer? I believe we could merge the first PR rather quickly.

To avoid the formatting-check error, you can execute `mvn formatter:format` (and perhaps check whether `mvn verify -DskipTests=True` results in any errors.

kshehata commented:
Sure thing! I've removed any externally visible changes that I can see and hidden it behind a command line flag. I've also fixed the formatting and PMD problems. Is this simple enough to move forward with, or is there more that needs to be deferred?

mrigger commented on src/sqlancer/FoundBugException.java:
```diff
@@ -0,0 +1,26 @@
+    private final Reproducer reproducer;
+
+    public interface Reproducer {
+        boolean bugStillTriggers(SQLite3GlobalState globalState);
```
Just noticed that this is SQLite-specific. Would it be possible to use generics here?

kshehata replied:
I'm trying to replace with the generics now but I'm getting all kinds of weird errors. I'll keep working on it, but for now let's assume this is fixed and review everything else.

kshehata replied:
Just realized a fatal problem in our design: [Throwables cannot be Generic](https://stackoverflow.com/questions/501277/why-doesnt-java-allow-generic-subclasses-of-throwable). As such, I don't think we can keep a reference to a generic Reproducer without throwing away the generic type info, which is problematic. I think we'll need to rethink how this all works if we're to make this properly generic, and probably move away from using Exceptions for finding bugs.

mrigger commented on src/sqlancer/ReducerExample.java:
```diff
@@ -0,0 +1,89 @@
+import sqlancer.common.query.Query;
+import sqlancer.sqlite3.SQLite3GlobalState;
+
+public class ReducerExample<G extends GlobalState<O, ?, C>, O extends DBMSSpecificOptions<?>, C extends SQLancerDBConnection>
```
Perhaps `SimpleReducer` or `StatementReducer` is a better name? No need to address this in this PR.

kshehata replied:
Done.

mrigger commented on src/sqlancer/sqlite3/gen/ddl/SQLite3IndexGenerator.java:
```diff
@@ -44,6 +44,7 @@ private SQLQueryAdapter create() throws SQLException {
         errors.add("non-deterministic use of date() in an index");
         errors.add("non-deterministic use of datetime() in an index");
         errors.add("The database file is locked");
+        errors.add("Abort due to constraint violation");
```
Is this added error intended? I suspect that this might be for an older or more recent version of SQLite that we are currently testing?

kshehata replied:
No, I think this was an artifact of a merge or something previous. I've removed it.

mrigger reviewed (commented):
Great, thanks! Could have a look at the comments (the one about the change in the index generator and SQLite-specific class in the `Reproducer` interface), and see if it is possible to address them easily or not? After that, we can merge the PR - please rebase on the latest master branch.

For code style violations in future PRs, it would be great if you could amend the changes to the commit that introduced them (see https://github.com/sqlancer/sqlancer/blob/master/CONTRIBUTING.md#commit-history).

kshehata commented:
> Great, thanks! Could have a look at the comments (the one about the change in the index generator and SQLite-specific class in the `Reproducer` interface), and see if it is possible to address them easily or not? After that, we can merge the PR - please rebase on the latest master branch.

I've rebased to master (BTW: consider [renaming to main](https://www.zdnet.com/article/github-to-replace-master-with-main-starting-next-month/) when you get a chance). I'm working on the generics now, but let's move ahead with the review to make sure there are no other issues that need to be addressed.

> For code style violations in future PRs, it would be great if you could amend the changes to the commit that introduced them (see https://github.com/sqlancer/sqlancer/blob/master/CONTRIBUTING.md#commit-history).

I'm not sure I understand. You'd like me to modify a previous commit to address a comment? But then how would you see what changes I've made since you last reviewed, wouldn't that make more work for you? Why not just squash the whole branch before merging?

mrigger commented:
> > Great, thanks! Could have a look at the comments (the one about the change in the index generator and SQLite-specific class in the `Reproducer` interface), and see if it is possible to address them easily or not? After that, we can merge the PR - please rebase on the latest master branch.
> 
> I've rebased to master (BTW: consider [renaming to main](https://www.zdnet.com/article/github-to-replace-master-with-main-starting-next-month/) when you get a chance). I'm working on the generics now, but let's move ahead with the review to make sure there are no other issues that need to be addressed.

Yes, performing the rename has been on my todo list for a while now. Should we merge the PR for now, or would you like to first address this properly?

> 
> > For code style violations in future PRs, it would be great if you could amend the changes to the commit that introduced them (see https://github.com/sqlancer/sqlancer/blob/master/CONTRIBUTING.md#commit-history).
> 
> I'm not sure I understand. You'd like me to modify a previous commit to address a comment? But then how would you see what changes I've made since you last reviewed, wouldn't that make more work for you? Why not just squash the whole branch before merging?

I was referring to the code style violations. Squashing as an alternative also sounds good.

mrigger reviewed (approved):
LGTM! Would you like that we merge the PR now, or after we found a way to address the generics issue? Either is fine with me.
